assets/js/linear_algebra/dot_product_collide.js

```diff
@@ -0,0 +1,344 @@
+let dot_product_collide = (function() {
+
+let origin = [300, 140], 
+    scale = 70, 
+    scatter = [], 
+    axis = [],
+    expectedAxis = [],
+    beta = 0, alpha = 0, 
+    startAngleX = Math.PI/8 * 2.65,
+    startAngleY = -Math.PI/8,
+    startAngleZ = Math.PI/8 * 0.6,
+    axis_len = 2,
+    unit = axis_len/10,
+    lib = null,
+    svg = null,
+    uTvv_opacity = 0;
+
+
+function select_svg(svg_id) {  
+  svg = d3.select(svg_id);
+
+  lib = space_plot_lib(
+    svg,
+    origin,
+    scale,
+    is_2d=false);
+
+  svg = svg.call(d3.drag()
+           .on('start', drag_start)
+           .on('drag', dragged)
+           .on('end', drag_end))
+           .append('g'); 
+}
+
+
+function plot(scatter, axis, tt){
+  let u = scatter[0],
+      v = scatter[1];
+
+  let lines = [], points = [];
+
+  lib.plot_lines(axis, tt);
+
+  scatter.forEach(function(d){
+    lines.push(...lib.create_segments(d));
+  })
+
+  basis = {
+      ex: axis[1/unit - 1 + axis_len/unit * 0][1], 
+      ey: axis[1/unit - 1 + axis_len/unit * 1][1], 
+      ez: axis[1/unit - 1 + axis_len/unit * 2][1],
+  };
+
+  let uTv = lib.dot_product(u, v);
+  
+  let uTvv = {
+      x: v.x * uTv,
+      y: v.y * uTv,
+      z: v.z * uTv,
+      r: 1.8,
+      color: 'grey'
+  }
+
+  let uTvv_line = [
+      {x: 0, y: 0, z: 0},
+      {x: uTvv.x, y: uTvv.y, z: uTvv.z,
+       tt: true}
+  ]
+  uTvv_line.color = 0
+  uTvv_line.centroid_z = 1000;
+
+  lines.push(uTvv_line);
+
+  lib.create_dash_segments(u, uTvv).forEach(
+      function(d) {
+        d.centroid_z = -900;
+        lines.push(d);
+      }
+  )
+
+  scatter.forEach(function(d, i){
+    let coord = lib.dot_basis(d, basis);
+    d.coord = coord;
+    let point = Object.assign({}, d);
+    if (i == 0) {
+      point.text = 'u = ';
+    } else {
+      point.text = 'v = ';
+    }
+    point.text += '['.concat(
+        coord.x.toFixed(2),
+        ', ',
+        coord.y.toFixed(2),
+        ', ',
+        coord.z.toFixed(2),
+        ']');
+    points.push(point);
+  })
+
+  points.push(uTvv);
+
+  lib.plot_lines(lines, tt, 'arrow');
+  lib.plot_points(points, tt,
+                  drag_point_fn=function(d, i){
+                    if ( i == 0) {
```

Comment:
rm the space between `(` and `i`.

Reply:
also check other *.js that has the same code as well.

Reply:
done